tracing-forest/src/printer/pretty.rs

```diff
@@ -0,0 +1,210 @@
+    }
+
+    fn format_indent(&self, indent: &mut Vec<Indent>, writer: &mut String) -> fmt::Result {
+        for indent in indent.iter() {
+            writer.write_str(indent.repr())?;
+        }
+        Ok(())
+    }
+
+    fn format_event(&self, event: &Event, writer: &mut String) -> fmt::Result {
+        let tag = event.tag();
+        let message = event.message().unwrap_or("");
+
+        write!(writer, "{} [{}]: {}", tag.icon(), tag, message)?;
+
+        for field in event.fields().iter() {
+            write!(writer, " | {}: {}", field.key(), field.value())?;
+        }
+
+        writeln!(writer)
+    }
+
+    fn format_span(
+        &self,
+        span: &Span,
+        duration_root: Option<f64>,
+        indent: &mut Vec<Indent>,
+        writer: &mut String,
+    ) -> fmt::Result {
+        let total_duration = span.total_duration().as_nanos() as f64;
+        let inner_duration = span.inner_duration().as_nanos() as f64;
+        let root_duration = duration_root.unwrap_or(total_duration);
+        let percent_total_of_root_duration = 100.0 * total_duration / root_duration;
+
+        write!(
+            writer,
+            "{} [ {} | ",
+            span.name(),
+            DurationDisplay(total_duration)
+        )?;
+
+        if inner_duration > 0.0 {
+            let base_duration = span.base_duration().as_nanos() as f64;
+            let percent_base_of_root_duration = 100.0 * base_duration / root_duration;
+            write!(writer, "{:.2}% / ", percent_base_of_root_duration)?;
+        }
+
+        writeln!(writer, "{:.2}% ]", percent_total_of_root_duration)?;
+
+        if let Some((last, remaining)) = span.children().split_last() {
+            match indent.last_mut() {
+                Some(edge @ Indent::Turn) => *edge = Indent::Null,
+                Some(edge @ Indent::Fork) => *edge = Indent::Line,
+                _ => {}
+            }
+
+            indent.push(Indent::Fork);
+
+            for tree in remaining {
+                if let Some(edge) = indent.last_mut() {
+                    *edge = Indent::Fork;
+                }
+                self.format_tree(tree, Some(root_duration), indent, writer)?;
+            }
+
+            if let Some(edge) = indent.last_mut() {
+                *edge = Indent::Turn;
+            }
+            self.format_tree(last, Some(root_duration), indent, writer)?;
+
+            indent.pop();
+        }
+
+        Ok(())
+    }
+}
+
+enum Indent {
+    Null,
+    Line,
+    Fork,
+    Turn,
+}
+
+impl Indent {
+    fn repr(&self) -> &'static str {
+        match self {
+            Self::Null => "   ",
+            Self::Line => "│  ",
+            Self::Fork => "┝━ ",
+            Self::Turn => "┕━ ",
+        }
+    }
+}
+
+struct DurationDisplay(f64);
+
+// Taken from chrono
+impl fmt::Display for DurationDisplay {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut t = self.0;
+        for unit in ["ns", "µs", "ms", "s"] {
```

Comment:
One extra comment from having looking at these logs, as I am not always a bright spark, I was confused between ns and us ^^;. It might be worth removing us since I think most developers we either look in "ms" or "ns". IE something like kanidm, it's useful to have everything is ms only so we can compare timings between all ops without needing to look too much into what units are used. So maybe remove us?

Reply:
If we removed µs then there would be almost no middle ground between nanoseconds and milliseconds. For cases where it would be like 150µs, do you propose we write 0.15ms? or 150000ns? If we were to round to 0ms, where do we draw the line between rounding up to ms and having a very long ns?

Reply:
Yeah, I'd do it as 0.15 ms :)

Reply:
Don't worry about this for now :)